README
NAME RDF::Simple
DESCRIPTION
        This package is for very simple manipulations
        of RDF/XML serialisations of RDF graphs. 
        It consists of:

        RDF::Simple::Parser
        RDF::Simple::Serialiser
    
        Please consult the individual pod for these packages.
        The parser requires XML::SAX,
        The serialiser requires Template Toolkit (Template)

        Also provided is RDF::Simple::NS,
        a utility class for XML namespaces in RDF

        The aim here is to keep things Simple: 
        e.g., the parser doesn't differentiate between
        literal and resource values in the model
        All you get back is a bucket-o-triples
        (array of arrays)
        The serialiser does its best to do DWYM.

        If you want a more complex and involved RDF API,
        I'd suggest looking at RDF::Core
        or at the redland RDF application toolkit
        at http://www.redland.opensource.ac.uk
    
        This is a beta release, and may well contain bugs.
        Please report them!
    
AUTHOR
        Jo Walsh <jo@london.pm.org>

LICENSE
        This package and its contents are available 
        under the same terms as perl itself



Changes
2009-07-03  Kingpin  <Martin@BLUEMAX>

	* lib/RDF/Simple/Serialiser.pm: don't require subject to be a full URI

	* lib/RDF/Simple/Parser/Handler.pm: Try to parse the xml:base from the header and use it when needed

2009-04-11  Kingpin  <Martin@BLUEMAX>

	* lib/RDF/Simple/Parser/Handler.pm (propertyElt): fixed to allow literals with value '0'

2009-04-09  Kingpin  <Martin@BLUEMAX>

	* lib/RDF/Simple/Parser/Sink.pm: got rid of this empty unused module

	* lib/RDF/Simple/Parser/Handler.pm (propertyElt): fixed wrong method name called!

	* lib/RDF/Simple/Parser.pm (parse_rdf): fixed to handle case when handler returns no result

2009-04-08  Kingpin  <Martin@BLUEMAX>

	* lib/RDF/Simple/Parser/Handler.pm (propertyElt): fixed to handle simple rdf:resource attributes

2009-03-26  Kingpin  <Martin@BLUEMAX>

	* lib/RDF/Simple/Serialiser.pm (_make_object): if object is scalar ref, output as literal (even if it looks like a URI)

2009-03-02  Kingpin  <Martin@BLUEMAX>

	* lib/RDF/Simple/Parser/Handler.pm (characters): FIXED to allow data to be parsed in multiple chunks

2009-03-01  Kingpin  <Martin@BLUEMAX>

	* lib/RDF/Simple/Parser/Handler.pm (emptyPropertyElt): FIXED for RT#43688: do not create bNode for empty-string property value

2008-12-06  Kingpin  <Martin@BLUEMAX>

	* simplified the tests

	* massive code clean ups (now uses Class::MethodMaker)

2008-10-05  Kingpin  <Martin@BLUEMAX>

	* lib/RDF/Simple/Serialiser.pm (make_object): fixed for RT ticket 29067
	(make_object): fixed for RT ticket 35484
	(make_object): fixed for RT ticket 37335

	* lib/RDF/Simple/Parser.pm (parseTypeCollectionPropertyElt): applied fix described in RT ticket 25302

	* lib/RDF/Simple/Serializer.pm: new file

	* Makefile.PL: converted to Module::Install

Revision history for Perl module RDF::Simple

0.3 18th April 2006
	- removed dependency on Template::Toolkit and Class::MethodMaker
	( latter build was breaking; former too complex.)

0.22 24th Sept 2005 
	- patch to recognise file:// and urn: as URIs
 
0.2 Mon May 24th 20:10:00 UTC
	- applied patch to deal with parsing URIs via a http_proxy
	- fixed a serialisato do with generated node ids
 
0.01 Tue Aug  5 23:30:00 2003
	- original version; created by ExtUtils::ModuleMaker 0.32




lib/RDF/Simple/Parser.pm

# $Id: Parser.pm,v 1.12 2009/05/01 03:30:12 Martin Exp $

use strict;
use warnings;

=head1 NAME

RDF::Simple::Parser - convert RDF string to bucket of triples

=head1 DESCRIPTION

A simple RDF/XML parser -
reads a string containing RDF in XML
returns a 'bucket-o-triples' (array of arrays)

=head1 SYNOPSIS

  my $uri = 'http://www.zooleika.org.uk/bio/foaf.rdf';
  my $rdf = LWP::Simple::get($uri);
  my $parser = RDF::Simple::Parser->new(base => $uri)
  my @triples = $parser->parse_rdf($rdf);
  # Returns an array of array references which are triples

=head1 METHODS

=over

=cut

package RDF::Simple::Parser;

use constant DEBUG => 0;

use File::Slurp;
use LWP::UserAgent;
use RDF::Simple::Parser::Handler;
use XML::SAX qw(Namespaces Validation);

our
$VERSION = do { my @r = (q$Revision: 1.12 $ =~ /\d+/g); sprintf "%d."."%03d" x $#r, @r };

# Use a hash to implement objects of this type:
use Class::MakeMethods::Standard::Hash (
                                        new => 'new',
                                        scalar => [ qw( base http_proxy )],
                                       );

=item new( [ base => 'http://example.com/foo.rdf' ])

Create a new RDF::Simple::Parser object.

'base' supplies a base URI for relative URIs found in the document

'http_proxy' optionally supplies the address of an http proxy server.
If this is not given, it will try to use the default environment settings.

=cut

=item parse_rdf($rdf)

Accepts a string which is an RDF/XML document
(complete XML, with headers)

Returns an array of array references which are RDF triples.

=cut

sub parse_rdf
  {
  my ($self, $rdf) = @_;
  DEBUG && print STDERR " DDD Parser::parse_rdf()\n";
  my $handler = new RDF::Simple::Parser::Handler(q{deprecated argument},
                                                 qnames => 1,
                                                 base => $self->base,
                                                );
  my $factory = new XML::SAX::ParserFactory;
  $factory->require_feature(Namespaces);
  my $parser = $factory->parser(Handler => $handler);
  $parser->parse_string($rdf);
  my $res = $handler->result;
  return $res ? @$res : $res;
  } # parse_rdf


=item parse_file($sFname)

Takes one argument, a string which is a fully qualified filename.
Reads the contents of that file,
parses it as RDF,
and returns the same thing as parse_rdf().

=cut

sub parse_file
  {
  my $self = shift;
  my $sFname = shift || return;
  my $sRDF = read_file($sFname) || return;
  return $self->parse_rdf($sRDF);
  } # parse_file


=item parse_uri($uri)

Accepts a string which is a fully qualified http:// uri
at which some valid RDF lives.
Fetches the remote file and returns the same thing as parse_rdf().

=cut

sub parse_uri
  {
  my $self = shift;
  my $uri = shift || return;
  my $rdf;
  eval
    {
    $rdf = $self->ua->get($uri)->content;
    };
  warn ($@) if $@;
  if ($rdf && ($rdf ne q{}))
    {
    # print STDERR " DDD will parse_rdf(===$rdf===)\n";
    $self->base($uri);
    return $self->parse_rdf($rdf);
    } # if
  } # parse_uri

sub ua
  {
  my $self = shift;
  unless ($self->{_ua}) {
    $self->{_ua} = LWP::UserAgent->new(timeout => 30);
    if ($self->http_proxy) {
      $self->{_ua}->proxy('http',$self->http_proxy);
      } else {
        $self->{_ua}->env_proxy;
        }
    }
  return $self->{_ua};
  } # ua

=back

=head1 BUGS

Please report bugs via the RT web site L<http://rt.cpan.org/Ticket/Create.html?Queue=RDF-Simple>

=head1 AUTHOR

Jo Walsh <jo@london.pm.org>
Currently maintained by Martin Thurn <mthurn@cpan.org>

=head1 LICENSE

This module is available under the same terms as perl itself

=cut

1;

__END__
